contracts/multisig-prover/src/encoding/abi2/mod.rs

```diff
@@ -0,0 +1,205 @@
+use alloy_primitives::Address;
+use alloy_sol_types::{sol, SolValue};
+use cosmwasm_std::{HexBinary, Uint128, Uint256};
+use k256::{elliptic_curve::sec1::ToEncodedPoint, PublicKey};
+use sha3::{Digest, Keccak256};
+
+use axelar_wasm_std::hash::Hash;
+use multisig::{key::PublicKey as MultisigPublicKey, msg::Signer, worker_set::WorkerSet};
+
+use crate::{error::ContractError, payload::Payload};
+
+sol!("src/encoding/abi2/solidity/AmplifierGatewayTypes.sol");
+sol!("src/encoding/abi2/solidity/WeightedMultisigTypes.sol");
+
+impl From<&Payload> for CommandType {
+    fn from(payload: &Payload) -> Self {
+        match payload {
+            Payload::Messages(_) => CommandType::ApproveMessages,
+            Payload::WorkerSet(_) => CommandType::RotateSigners,
+        }
+    }
+}
+
+impl WeightedSigners {
+    pub fn hash(&self) -> Hash {
+        Keccak256::digest(self.abi_encode()).into()
+    }
+}
+
+impl From<&Signer> for WeightedSigner {
+    fn from(signer: &Signer) -> Self {
+        let weight: Uint128 = signer
+            .weight
+            .try_into()
+            .expect("weight is too large to convert to Uint128");
+
+        WeightedSigner {
+            signer: evm_address(&signer.pub_key).expect("failed to convert pub key to evm address"),
+            weight: weight.u128(),
+        }
+    }
+}
+
+impl From<&WorkerSet> for WeightedSigners {
+    fn from(worker_set: &WorkerSet) -> Self {
+        let mut signers = worker_set
+            .signers
+            .values()
+            .map(WeightedSigner::from)
+            .collect::<Vec<_>>();
+
+        signers.sort_by_key(|weighted_signer| weighted_signer.signer);
+
+        let threshold: u128 = Uint128::try_from(worker_set.threshold)
+            .expect("threshold is too large to convert to Uint128")
+            .u128();
+
+        WeightedSigners {
+            signers,
+            threshold,
+            nonce: Uint256::from(worker_set.created_at).to_le_bytes().into(),
+        }
+    }
+}
+
+pub fn payload_hash_to_sign(
+    domain_separator: &Hash,
+    signer: &WorkerSet,
+    payload: &Payload,
+) -> HexBinary {
+    let signer = WeightedSigners::from(signer);
+    let data_to_sign = encode(payload);
+
+    // Prefix for standard EVM signed data https://eips.ethereum.org/EIPS/eip-191
+    let unsigned = [
+        "\x19Ethereum Signed Message:\n96".as_bytes(), // 96 is the length of the trailing bytes
```

Comment:
we should either compute the length of the trailing bytes, or at least have a test that 96 is the length

Reply:
keeping consistence with [solidity](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/v5.8.0/contracts/governance/BaseWeightedMultisig.sol#L186), the function is tested in unit test. 

a bit hesitate to compute length because clippy complains for plain arithmetic operation, use wrapper type leads to unnecessary complexity